File: unique_binary_search_trees.py
'''
    Mathematics: https://en.wikipedia.org/wiki/Catalan_number

    Given an integer n, return the number of structurally unique BST's (binary search trees)
    which has exactly n nodes of unique values from 1 to n.

    Example 1:

    Input: n = 3
    Output: 5

    Example 2:

    Input: n = 1
    Output: 1
'''


def catalan(number):
    if number <= 1:
        return 1

    nodes = [0] * (number+1)

    nodes[0] = 1
    nodes[1] = 1

    for i in range(2, number+1):
        nodes[i] = 0
        for j in range(0, i):
            nodes[i] += nodes[j] * nodes[i-j-1]

    return nodes[number]


class Solution:
    def numTrees(self, n: int) -> int:
        return catalan(n)

File: Coding Practice/uniqueBinarySearchTrees.java
/*
Given A, generate all structurally unique BST's (binary search trees) that store values 1...A.


Problem Constraints
1 <= A <= 15


Input Format
The first argument is an integer A.


Output Format
Return an array of TreeNode


Example Input
A = 3


Example Output
   1         3     3      2      1
    \       /     /      / \      \
     3     2     1      1   3      2
    /     /       \                 \
   2     1         2                 3

*/

//Definition for binary tree
 class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
 }

public class Solution {
	public ArrayList<TreeNode> generateTrees(int a) {
        return trees(1, a);
    }
    
    private ArrayList<TreeNode> trees(int start, int end) {
        ArrayList<TreeNode> nodes = new ArrayList<>();
        
        if(start > end) {
            nodes.add(null);
            return nodes;
        }
        
        for(int i=start; i <= end; i++) {
            ArrayList<TreeNode> left = trees(start, i-1);
            ArrayList<TreeNode> right = trees(i+1, end);
            
            for(TreeNode l : left) {
                for(TreeNode r : right) {
                    TreeNode root = new TreeNode(i);
                    root.left = l;
                    root.right = r;
                    nodes.add(root);
                }
            }
        }
        
        return nodes;
    }       
}